File: apps/mobile-flutter/android/app/src/main/kotlin/com/voice2text/app/transcription/TranscriptionResult.kt
package com.voice2text.app.transcription

data class TranscriptionSegmentResult(
    val sequenceId: Int,
    val text: String,
    val startMs: Long,
    val endMs: Long,
    val isFinal: Boolean = true,
    val source: String = "standard_offline",
    val confidence: Double? = null,
) {
    fun toMap(): Map<String, Any?> =
        mapOf(
            "sequenceId" to sequenceId,
            "text" to text,
            "startMs" to startMs,
            "endMs" to endMs,
            "isFinal" to isFinal,
            "source" to source,
            "confidence" to confidence,
        )
}

data class TranscriptionResult(
    val mergedText: String,
    val segments: List<TranscriptionSegmentResult>,
) {
    init {
        require(segments.isNotEmpty()) { "转写结果不包含可持久化片段" }
        var previousEndMs = 0L
        segments.forEachIndexed { index, segment ->
            require(segment.sequenceId == index) { "转写片段序号不连续" }
            require(segment.text.isNotBlank()) { "转写片段文本为空" }
            require(segment.startMs >= previousEndMs) { "转写片段时间范围重叠" }
            require(segment.endMs > segment.startMs) { "转写片段时间范围无效" }
            require(segment.isFinal) { "生产转写结果只能包含最终片段" }
            require(segment.source.isNotBlank()) { "转写片段来源为空" }
            require(segment.confidence == null || segment.confidence in 0.0..1.0) {
                "转写片段置信度无效"
            }
            previousEndMs = segment.endMs
        }
        require(mergedText.trim() == segments.joinToString(" ") { it.text.trim() }) {
            "合并文本与片段顺序不一致"
        }
    }

    fun toMap(): Map<String, Any?> =
        mapOf(
            "mergedText" to mergedText.trim(),
            "segments" to segments.map(TranscriptionSegmentResult::toMap),
        )

    companion object {
        fun fromSegments(segments: List<TranscriptionSegmentResult>): TranscriptionResult =
            TranscriptionResult(
                mergedText = segments.joinToString(" ") { it.text.trim() },
                segments = segments,
            )

        fun singleText(
            text: String,
            durationMs: Int,
        ): TranscriptionResult {
            val normalized = text.trim()
            return fromSegments(
                listOf(
                    TranscriptionSegmentResult(
                        sequenceId = 0,
                        text = normalized,
                        startMs = 0,
                        endMs = durationMs.coerceAtLeast(1).toLong(),
                    ),
                ),
            )
        }
    }
}

internal class VadTimestampMapper(
    private val sampleRate: Int,
) {
    private var previousEndMs = 0L

    init {
        require(sampleRate > 0)
    }

    fun map(
        sequenceId: Int,
        text: String,
        startSample: Long,
        sampleCount: Int,
        confidence: Double? = null,
    ): TranscriptionSegmentResult {
        require(startSample >= 0)
        require(sampleCount > 0)
        val rawStartMs = samplesToMs(startSample)
        val rawEndMs = samplesToMs(startSample + sampleCount)
        val startMs = maxOf(rawStartMs, previousEndMs)
        val endMs = maxOf(rawEndMs, startMs + 1)
        previousEndMs = endMs
        return TranscriptionSegmentResult(
            sequenceId = sequenceId,
            text = text.trim(),
            startMs = startMs,
            endMs = endMs,
            confidence = confidence,
        )
    }

    private fun samplesToMs(samples: Long): Long = (samples * 1000L) / sampleRate
}

File: apps/mobile-flutter/lib/features/transcription/model/transcription_result.dart
class TranscriptionSegmentResult {
  const TranscriptionSegmentResult({
    required this.sequenceId,
    required this.text,
    required this.startMs,
    required this.endMs,
    this.isFinal = true,
    this.source = 'standard_offline',
    this.confidence,
  });

  factory TranscriptionSegmentResult.fromMap(Map<Object?, Object?> map) {
    return TranscriptionSegmentResult(
      sequenceId: (map['sequenceId'] as num?)?.toInt() ?? -1,
      text: map['text'] as String? ?? '',
      startMs: (map['startMs'] as num?)?.toInt() ?? -1,
      endMs: (map['endMs'] as num?)?.toInt() ?? -1,
      isFinal: map['isFinal'] as bool? ?? true,
      source: map['source'] as String? ?? 'standard_offline',
      confidence: (map['confidence'] as num?)?.toDouble(),
    );
  }

  final int sequenceId;
  final String text;
  final int startMs;
  final int endMs;
  final bool isFinal;
  final String source;
  final double? confidence;

  Map<String, Object?> toMap() => <String, Object?>{
    'sequenceId': sequenceId,
    'text': text,
    'startMs': startMs,
    'endMs': endMs,
    'isFinal': isFinal,
    'source': source,
    'confidence': confidence,
  };
}

class TranscriptionResult {
  TranscriptionResult({
    required String mergedText,
    required List<TranscriptionSegmentResult> segments,
  }) : mergedText = mergedText.trim(),
       segments = List<TranscriptionSegmentResult>.unmodifiable(segments) {
    validate();
  }

  factory TranscriptionResult.fromMap(Map<Object?, Object?> map) {
    final rawSegments = map['segments'];
    return TranscriptionResult(
      mergedText: map['mergedText'] as String? ?? '',
      segments: rawSegments is List
          ? rawSegments
                .whereType<Map>()
                .map(
                  (segment) => TranscriptionSegmentResult.fromMap(
                    Map<Object?, Object?>.from(segment),
                  ),
                )
                .toList(growable: false)
          : const <TranscriptionSegmentResult>[],
    );
  }

  factory TranscriptionResult.singleText(
    String text, {
    required int durationMs,
  }) {
    final normalized = text.trim();
    return TranscriptionResult(
      mergedText: normalized,
      segments: <TranscriptionSegmentResult>[
        TranscriptionSegmentResult(
          sequenceId: 0,
          text: normalized,
          startMs: 0,
          endMs: durationMs <= 0 ? 1 : durationMs,
        ),
      ],
    );
  }

  final String mergedText;
  final List<TranscriptionSegmentResult> segments;

  Map<String, Object?> toMap() => <String, Object?>{
    'mergedText': mergedText,
    'segments': segments.map((segment) => segment.toMap()).toList(),
  };

  void validate() {
    if (segments.isEmpty) {
      throw const FormatException('转写结果不包含可持久化片段');
    }
    var previousEndMs = 0;
    final texts = <String>[];
    for (var index = 0; index < segments.length; index += 1) {
      final segment = segments[index];
      final normalizedText = segment.text.trim();
      if (segment.sequenceId != index ||
          normalizedText.isEmpty ||
          segment.startMs < previousEndMs ||
          segment.endMs <= segment.startMs ||
          !segment.isFinal ||
          segment.source.trim().isEmpty ||
          (segment.confidence != null &&
              (segment.confidence! < 0 || segment.confidence! > 1))) {
        throw FormatException('第 $index 个转写片段无效');
      }
      previousEndMs = segment.endMs;
      texts.add(normalizedText);
    }
    if (mergedText != texts.join(' ')) {
      throw const FormatException('合并文本与最终片段顺序不一致');
    }
  }
}
